CMakeLists.txt

```diff
@@ -778,45 +778,82 @@ function(add_library_variants)
+          string(APPEND multilib_variants "  - ${flag}\n")
+        endforeach()
+
+        string(APPEND multilib_variants "  PrintOptions:\n")
+        string(REPLACE " " ";" multilib_printoptions_list ${compile_flags})
+        foreach(option ${multilib_printoptions_list})
+          string(APPEND multilib_variants "  - ${option}\n")
+        endforeach()
+
         install(
             DIRECTORY "${LLVM_BINARY_DIR}/${directory}/"
             DESTINATION "${directory}"
             COMPONENT llvm-toolchain
         )
     endwhile()
+    set(multilib_variants "${multilib_variants}" PARENT_SCOPE)
 endfunction()
 
+set(multilib_variants "")
+
 # Define which library variants to build and which flags to use.
-# The order is <parent dir> <arch> <name suffix> <flags> <qemu params>
+# The order is <arch> <name suffix> <compile flags> <multilib selection flags> <qemu params>
 add_library_variants(
-    aarch64         ""                  "-march=armv8-a"                                        "-M raspi3b"
-    armv4t          ""                  "-march=armv4t"                                         "-M musicpal -cpu arm926"
-    armv5te         ""                  "-march=armv5te"                                        "-M musicpal -cpu arm926"
-    armv6m          soft_nofp           "-mfloat-abi=soft -march=armv6m"                        "-M microbit"
-    armv7m          soft_nofp           "-mfloat-abi=soft -march=armv7m+nofp"                   "-M mps2-an385 -cpu cortex-m3"
-    armv7em         soft_nofp           "-mfloat-abi=soft -march=armv7em -mfpu=none"            "-M mps2-an386 -cpu cortex-m4"
-    armv7em         hard_fpv4_sp_d16    "-mfloat-abi=hard -march=armv7em -mfpu=fpv4-sp-d16"     "-M mps2-an386 -cpu cortex-m4"
-    armv7em         hard_fpv5_d16       "-mfloat-abi=hard -march=armv7em -mfpu=fpv5-d16"        "-M mps2-an500 -cpu cortex-m7"
-    armv8m.main     soft_nofp           "-mfloat-abi=soft -march=armv8m.main+nofp"              "-M mps2-an505 -cpu cortex-m33"
-    armv8m.main     hard_fp             "-mfloat-abi=hard -march=armv8m.main+fp"                "-M mps3-an524 -cpu cortex-m33"
-    armv8.1m.main   soft_nofp_nomve     "-mfloat-abi=soft -march=armv8.1m.main+nofp+nomve"      "-M mps3-an547 -cpu cortex-m55"
-    armv8.1m.main   hard_fp             "-mfloat-abi=hard -march=armv8.1m.main+fp"              "-M mps3-an547 -cpu cortex-m55"
-    armv8.1m.main   hard_nofp_mve       "-mfloat-abi=hard -march=armv8.1m.main+nofp+mve"        "-M mps3-an547 -cpu cortex-m55"
+    aarch64         ""                  "-march=armv8-a"                                        "target=armv8a-none-unknown-eabi"                                                                          "-M raspi3b"
+    armv4t          ""                  "-march=armv4t"                                         "target=armv4t-none-unknown-eabi mfpu=none"                                                                "-M musicpal -cpu arm926"
```

Comment:
If `mfpu=<anything>` is compatible with `mfpu=none` then what's the value in specifying `mfpu=none` as a requirement?

Reply:
You're right that those are inconsistent, but (documenting here our discussion in person this morning) I think actually the claim of compatibility is the one I should change. For normal PCS purposes it's safe to call a non-FPU-using function from an FPU-using one (at least as long as the soft-float ABI is in use in both cases), but in outlying cases like `setjmp` and C++ exceptions, it matters that the library knows that FP registers are present, because it needs to save and restore them.

So I'm changing the FlagMap to stop claiming that `mfpu=none` is compatible with anything that has 16 d-registers.